### ISSUE TITLE
ikos-analyzer stuck and/or running out of memory

### ISSUE THREAD

**Issue by victorjulien** [NASA-SW-VnV/ikos#49]
When analyzing the suricata code base (~300k lines of C), the ikos-analyzer uses excessive memory and appears to get stuck in some kind of loop.

<pre>
Analyze src/suricata? [Y/n]
[*] Running ikos src/suricata.bc -o src/suricata.db
[*] Running ikos preprocessor
[*] Running ikos analyzer
[*] Translating LLVM bitcode to AR
[*] Running liveness analysis
[*] Running fixpoint profile analysis
[*] Running interprocedural value analysis
[*] Analyzing entry point: main
</pre>

I killed it at 30G of memory use to avoid the OOM killer kicking in and disrupting other processes on this box.

The way I set up this test:
<pre>
docker run -it ubuntu:18.10 /bin/bash
apt update && apt upgrade -y && apt -y install clang-7 libpcre3 libpcre3-dbg libpcre3-dev build-essential autoconf automake libtool libpcap-dev libnet1-dev libyaml-0-2 libyaml-dev pkg-config zlib1g zlib1g-dev libcap-ng-dev libcap-ng0 make libmagic-dev libjansson-dev libboost-dev libboost-atomic1.67-dev libboost-date-time1.67-dev libboost-iostreams1.67-dev libboost-graph1.67-dev libboost-filesystem1.67-dev libboost-numpy1.67-dev libboost-python1.67-dev libboost-random1.67-dev libboost-thread1.67-dev libboost1.67-tools-dev libboost-test1.67-dev build-essential git cmake libgmp-dev libsqlite3-dev libapron-dev

git clone https://github.com/NASA-SW-VnV/ikos
cd ikos/
mkdir build
cd build/
cmake -DLLVM_CONFIG_EXECUTABLE=/usr/bin/llvm-config-7 ../
make
make install
export PATH=/ikos/install/bin/:$PATH
cd /
git clone https://github.com/OISF/suricata
cd suricata/
git clone https://github.com/OISF/libhtp
bash autogen.sh 
ikos-scan ./configure --disable-shared --disable-gccmarch-native
ikos-scan make
</pre>

When running ikos with verbose output, I noticed that some functions were analyzed a lot of times. I'm not sure if that is normal.
<pre>
# ikos -v src/suricata.bc &> debug.log
... killed when process uses 10G of mem ...
# cat debug.log | sort | uniq -c | sort -nr |head -n25
  11929 [.] Analyzing function: SCMapEnumValueToName
  11928 [.] Analyzing function: SCLocalTime
   5868 [.] Analyzing function: SCErrorToString
   3003 [.] Analyzing function: strlcpy
   2982 [.] Analyzing function: SCLogMessageJSON
   2982 [.] Analyzing function: SCLogMessageGetBuffer
   2982 [.] Analyzing function: CreateIsoTimeString
   2981 [.] Analyzing function: json_decref.8298
   2930 [.] Analyzing function: SCLogMatchFGFilter
   1987 [.] Analyzing function: SCLogPrintToStream
   1465 [.] Analyzing function: SCLogMatchFGFilterWL
   1465 [.] Analyzing function: SCLogMatchFGFilterBL
   1465 [.] Analyzing function: SCLogMatchFDFilter
    994 [.] Analyzing function: SCLogReopen
    994 [.] Analyzing function: SCLogPrintToSyslog
    994 [.] Analyzing function: SCLogMapLogLevelToSyslogLevel
    514 [.] Analyzing function: SCLogMessage
     45 [.] Analyzing function: ConfNodeLookupChild
     43 [.] Analyzing function: ConfNodeNew
     29 [.] Analyzing function: RunModeRegisterNewRunMode
     29 [.] Analyzing function: RunModeGetCustomMode
     14 [.] Analyzing function: ConfSetFinal
     14 [.] Analyzing function: ConfNodeFree
     14 [.] Analyzing function: ConfGetNodeOrCreate
      8 [.] Analyzing function: LiveRegisterDeviceName
</pre>

Some of the functions in this list are called from a lot of locations in the code (sometimes through macros), not sure if that is relevant. Not all of them are though. E.g. SCLogReopen is just directly called from one location, but that is from SCLogMessage, which is called for a lot of locations.

**Comment by victorjulien** [comment]
I've let it run just to see what happens, and ultimately the OOM-killer killed it on this 64G machine.

**Comment by practicalswift** [comment]
FWIW I've also seen cases where the `ikos` process appears to never complete. Let me know if examples are needed. I've tried to reduce a test case for this but reducing timeout test cases is tricky for obvious reasons :-)

**Comment by arthaud** [comment]
Hi @victorjulien and @practicalswift,

**Regarding the verbose output:**
Seeing the same function analyzed over and over again is normal. This is related to how the analysis works. Let's say we have a function `f`:
```c
void f() {
  int sum = 0;
  for(int i = 0; i < 10; i++) {
    sum += g(i);
  }
  return sum;
}
```
When the analysis reaches `f`, it computes a fixpoint to infer properties about the code. For that, it will dive into the loop for a few iterations (probably 3 or 4). During the analysis of the loop, it finds the call to `g` and starts analyzing `g`. So you will see the message `Analyzing function: g` 3 or 4 times. Now, if `g` also has a loop calling `h`, it will show `Analyzing function: h` 9 to 16 times, etc.

I would like to have a better status message, showing the current call stack. It would make it easier to understand how far the analysis is while being less verbose. But I didn't find the time to implement this.

**Regarding the memory usage:**
This is pretty worrying. Using up to 64Go is definitely not normal..

We should first make sure there is no memory leaks in IKOS. Have you tried to run it with valgrind?

If it's not due to a memory leak, I have an idea that could help. For each function `f`, we keep in memory the fixpoints of functions called by `f`. This makes the analysis faster since we can reuse the fixpoint on callees when we reached the fixpoint on `f`. If the memory usage is a concern, we can disable this "cache". The counterpart is that it will make the analysis slower. I'm thinking about adding an option `--disable-call-cache` or something (I need a better name..). In the meantime, I made a patch: [disable-call-cache.patch.zip](https://github.com/NASA-SW-VnV/ikos/files/2700790/disable-call-cache.patch.zip). Please try it and tell me if it fixes the issue.

**Regarding the case when ikos never completes:**
In theory, this should not happen. Of course, it could be a bug (see #7 fixed a few weeks ago).
If you can reproduce with a simple example, I will fix it.
If you cannot reproduce with a small example, maybe IKOS is just being slow.
I could try to figure it out if you give me the code to analyze, but it will take me a lot of time.

**Comment by arthaud** [comment]
Any feedback about my patch? Did it fix the issues with the memory usage?

**Comment by victorjulien** [comment]
I've tried the ikos git master, and it seems the behavior has changed. I'm not longer seeing the memory use, but I'm not confident it's because of a fix. Instead, it seems that the 'Translating LLVM bitcode to AR' gets stuck. I never see any output after that line, not even while adding -v or -vv to ikos.

Is there a way to figure out what it is doing? The longest I've let it run is about 1.5hrs, of which it spent almost all of its time in 'Translating LLVM bitcode to AR'. Memory remained stable.

**Comment by arthaud** [comment]
It looks like a bug in the translation from LLVM to AR.
It might be introduced by commit 945554845ea8981c4e5fe589115a760f413bb19b. Could you try to analyze your code with commit b03de9713ba8718d49e37a61b522a18229f70f64?

By the way, I updated the patch: [disable-call-cache.patch.zip](https://github.com/NASA-SW-VnV/ikos/files/2814371/disable-call-cache.patch.zip)

**Comment by victorjulien** [comment]
b03de9713ba8718d49e37a61b522a18229f70f64 gets past the "Translating LLVM bitcode to AR", with and without the patch.

With the patch the memory growth stops around ~800MiB, so that seems reasonable as well.

It's been running for 3 hours now, and I can't tell if that is normal or that perhaps it is stuck. I'll leave it running for now. It's using a 100% CPU on one core and sits at:
<pre>
Analyze src/suricata? [Y/n] y
[*] Running ikos src/suricata.bc -o src/suricata.db
[*] Running ikos preprocessor
[*] Running ikos analyzer
[*] Translating LLVM bitcode to AR
[*] Running liveness analysis
[*] Running fixpoint profile analysis
[*] Running interprocedural value analysis
[*] Analyzing entry point 'main'
</pre>

Does this runtime seem reasonable or do you think it's stuck? If so, any option I can try or analysis I can do?

**Comment by arthaud** [comment]
> [b03de97](https://github.com/NASA-SW-VnV/ikos/commit/b03de9713ba8718d49e37a61b522a18229f70f64) gets past the "Translating LLVM bitcode to AR", with and without the patch.

I will try to reproduce the bug and fix it then.

> Does this runtime seem reasonable or do you think it's stuck? If so, any option I can try or analysis I can do?

I think this is reasonable for such a huge code. For instance, the analysis of a huge C++ code took me 7 hours.

**Comment by arthaud** [comment]
The bug is fixed with 5f8f1bffc7551865967d23bc7ab2c462a67ba5c8.

I improved the translation of LLVM types to AR types, it should be faster now.
On my laptop, the translation from LLVM to AR for suricata is 12sec instead of 30sec.

**Comment by victorjulien** [comment]
Running the git master + your patch now. Just the git master would still show the memory explosion issue.

**Comment by victorjulien** [comment]
It's still running w/o any new output. Beginning to think that it's somehow stuck. Is there a way to find out?

**Comment by arthaud** [comment]
Unfortunately, there is no way to check that right now.

I would like to implement a progress status. It would show the current stack trace and the number of iterations on each loop. It would make it easy to spot an infinite loop, since IKOS should never do more than ~5 iterations for a given loop.

Unfortunately, I have another task to do, so I don't know when I will have time to implement this..

**Comment by victorjulien** [comment]
I've killed it after 6 days. Please let me know if you want me to test something again.

Also, this long run required your patch otherwise it would run out of memory in minutes.

**Comment by arthaud** [comment]
Using bbd015e11800716094ec68d5520b3cbb9968ae7f, you can now use the option `--no-fixpoint-cache` instead of my patch.

I will soon work on improving the logs of the analysis progress. This way, we should notice if the analyzer gets stuck.

**Comment by arthaud** [comment]
Using 18c3fef31a96c138e80fe19a71abcfaefe78e2aa, you can now see the progress of the analysis.

It will print the current stack frame every 2 seconds. It also shows the number of iterations done for each cycle, and indicate an increasing (↑) or decreasing (↓) iteration.

For instance, here is the progress of the analysis on a single loop, with comments:
```
test.c:10:2: Analyzing loop [iteration 1↑] (increasing iteration 1)
test.c:10:2: Analyzing loop [iteration 2↑] (increasing iteration 2)
test.c:10:2: Analyzing loop [iteration 3↑] (increasing iteration 3, start widening)
test.c:10:2: Analyzing loop [iteration 4↑] (increasing iteration 4, optional)
test.c:10:2: Analyzing loop [iteration 1↓] (decreasing iteration 1, fixpoint reached, start narrowing)
test.c:10:2: Analyzing loop [iteration 2↓] (decreasing iteration 2, optional)
(post fixpoint reached, analysis is done)
```

If the number of iterations for a given loop goes beyond 10, it is a bug. In that case, please open an issue.

**Comment by victorjulien** [comment]
hi @arthaud, it looks like you have fixed the issue. The current master works well for Suricata without any extra options needed:
```
Analyze src/suricata? [Y/n] 
[*] Running ikos src/suricata.bc -o src/suricata.db
[*] Running ikos preprocessor
[*] Running ikos analyzer
[*] Translating LLVM bitcode to AR
[*] Running liveness analysis
[*] Running fixpoint profile analysis
[*] Running interprocedural value analysis
[*] Analyzing entry point 'main'
[*] Checking properties for entry point 'main'

# Time stats:
ikos-analyzer: 19.909 sec
ikos-pp      : 5.457 sec
```

So I think this ticket can be closed. Thanks a lot for your help. Now I'll have to review the actual output!

**Comment by victorjulien** [comment]
I may have spoken too soon. It seems the reason it ran quickly is that it encountered 2 'error' conditions. I've fixed/workedaround those, and now runtime and memory usage is back to my initial report. Will test with the new option next.

**Comment by arthaud** [comment]
Any news about this?

**Comment by victorjulien** [comment]
I've started another run. This time on a Ubuntu 19.04 image with LLVM 8 and the ikos git master as of this morning.

Without the ```--no-fixpoint-cache``` option it still runs out of memory within a couple of minutes. With the option it remains within reasonable bounds wrt memory use. I have no idea about the run time yet.

I think if you consider the ```--no-fixpoint-cache``` option a solution to this ticket then the ticket can be closed.

**Comment by arthaud** [comment]
I doubt we will have a better solution than `--no-fixpoint-cache`, at least in the short term.

I will close this for now, but feel free to open a new ticket about similar issues.

**Comment by victorjulien** [comment]
Thanks @arthaud !

**Comment by phoon0416** [comment]
hi @arthaud 

i have some fixpoint cache related questions about the implementation of IKOS:
1. As far as i know, the fixpoint cache is not used during data flow analysis, but 
used in run_checkers stage, that is, we cannot accelerate the data flow analysis, 
we just accelerate run_checker
2. if we do not use fixpoint cache, why we cannot run_checks during the data 
flow analysis, for example, if function f call function g in a loop, it will iteratively 
analyze g for several times, say three times, when analyze funciton g the first 
time, it will generate a function fixpoint for g, we will store pre abstract value and  
post abstract value for each statements, not just pre and post for each basic block, 
and we will run run_checkers for function fixpoint of g, after this, this function 
fixpoint can be released, when we analyze g the second time or the third time, 
we can do the same work as the first time. although it may generate repetitive 
results for these three times, since the abstract value of funciton fixpoint of g 
of previous two times is a subset of the last time, but we don't need fixpoint cache, 
which will save the ikos from being kill for OOM when analyzing large scale projects.

Thanks

**Comment by arthaud** [comment]
1. Yes, the cache is only used when we converged (since otherwise it would be invalid). This is usually the last iteration of the fixpoint and also when we run checkers. It can speed up the analysis a little but don't expect huge wins.
2. If you store the pre and post invariant for each statement, that will likely use more memory than what we currently do (we only store invariants on basic block entries/exits). Hence this will not help you from being killed by OOM. Furthermore, we should only apply checkers when after we converged, otherwise we will have duplicate or event inconsistent error messages (we could get an error at iteration 0 which becomes a warning at iteration 1).
